### PR TITLE
[Workspace] add validation for data source in get and bulk_get methods

### PR DESCRIPTION
**Pull request by raintygao**
### Description

Add validation for data source in get and bulk get

## Screenshot

No UX updates

## Testing the changes

In workspace, when getting a data source that doesn't belong to this workspace, this request will be forbidden. Before we didn't do this validation.

## Changelog

- fix: add validation for data source in get and bulk_get methods

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 1]
```diff
@@ -179,6 +189,22 @@ export class WorkspaceSavedObjectsClientWrapper {
     return hasPermission;
   }
 
+  // Data source is a workspace level object, validate if data source has the permission to the given `workspaceIds`
+  private validateDataSourcePermissions = (
+    object: SavedObject,
+    request: OpenSearchDashboardsRequest
+  ) => {
+    const requestWorkspaceId = getWorkspaceState(request).requestWorkspaceId;
+    if (
```
Comment (wanglam) [inline src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: Nit: Shall we rewrite to below code?
```
return !requestWorkspaceId || !!object.workspaces?.includes(requestWorkspaceId)
```
This looks more clearable to me.

**Review by wanglam — APPROVED** [review]

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 1 above)
Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: Do you mean we use this code to replace ` requestWorkspaceId &&
      !(object.workspaces && object.workspaces?.includes(requestWorkspaceId))`? I suppose their effect would not be the same.

Comment (wanglam) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: Not replace the if condition. We can replace the whole if statement.
```ts
if (
      requestWorkspaceId &&
      !(object.workspaces && object.workspaces?.includes(requestWorkspaceId))
    ) {
      return false;
    } else {
      return true;
    }
```
can be refactor to
```ts
return !requestWorkspaceId || !!object.workspaces?.includes(requestWorkspaceId)
```
It has better readability to me. The method will return true when request workspace id is empty or object's workspaces attributes includes request workspace id.

Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: `!!` is used for converting undefined to false, right?  This looks more clear, thanks.

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 2]
```diff
@@ -179,6 +189,15 @@ export class WorkspaceSavedObjectsClientWrapper {
     return hasPermission;
   }
 
+  // Data source is a workspace level object, validate if data source has the permission to the given `workspaceIds`
```
Comment (ruanyl) [inline src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:192]:
```
validate if data source has the permission to the given `workspaceIds`
```
what do you mean by this? should it be
```
validate if the request have access to the data source within the requested workspace?
```

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts [hunk 3]
```diff
@@ -179,6 +189,15 @@ export class WorkspaceSavedObjectsClientWrapper {
...
+  private validateDataSourcePermissions = (
+    object: SavedObject,
+    request: OpenSearchDashboardsRequest
+  ) => {
+    const requestWorkspaceId = getWorkspaceState(request).requestWorkspaceId;
+    return !requestWorkspaceId || !!object.workspaces?.includes(requestWorkspaceId);
```
Comment (ruanyl) [inline src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: `!requestWorkspaceId` this covers the case when outside of a workspace?

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 2 above)
Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts]: Yes, updated.

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 3 above)
Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: No, in `WorkspaceState` type, `requestWorkspaceId` could be undefined. Here just want to ensure that the workspaceId to be compared is existing.

Comment (wanglam) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: The `requestWorkspaceId` should be exists if user in a workspace. In another words,  user should be out of workspace if requestWorkspaceId is undefined right?

Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: We don't need to care whether is in or out of workspace especially here.

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: Imaging user is in global(requestWorkspaceId is undefined) and try to get a datasource(without workspaces) by `id`, we should prevent user from getting the datasource detail.

Comment (raintygao) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: I think this may be a common case in OSD, this doesn't belong to the concept of workspace/assign/unassign, we should not do extra limitation.

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: @wanglam @ruanyl what's your opinion?
IMO, regarding the restriction will only be turned on when permission control and workspace is enabled, `Reading` it from all kinds of unexpected approaches should be prevented.

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: Yeah, agreed. There may be some use cases that users may access data source detail in global.

Comment (SuZhou-Joe) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts:198]: But can we add a unit test to cover the case when requestWorkspaceId is empty?

File: src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts (hunk 1 above)
Comment (ruanyl) [reply src/plugins/workspace/server/saved_objects/workspace_saved_objects_client_wrapper.ts]: Not a big concern, but I do like your original implementation, it's more readable to me, when I read the code, I can easily get that we only disallow the access to data source when request with workspace but the data source is not associated with the workspace. ;D

**Review by SuZhou-Joe — APPROVED** [review]

**Review by ruanyl — APPROVED** [review]
